cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -1416,6 +1416,120 @@ private int moveLeadershipInBatch() throws InterruptedException, ExecutionExcept
       return finishedTasks;
     }
 
+    private void handleProgressWithTopicDeletion(ExecutionTask task, List<ExecutionTask> finishedTasks, Set<Long> deletedTaskIds) {
+      // Handle topic deletion during the execution.
+      LOG.debug("Task {} is marked as finished because the topic has been deleted.", task);
+      finishedTasks.add(task);
+      deletedTaskIds.add(task.executionId());
+      _executionTaskManager.markTaskAborting(task);
+      _executionTaskManager.markTaskDone(task);
+    }
+
+    private void handleProgressWithCompletion(ExecutionTask task, List<ExecutionTask> finishedTasks) {
+      // Check to see if the task is done.
+      finishedTasks.add(task);
+      _executionTaskManager.markTaskDone(task);
+    }
+
+    /**
+     * Periodically checks the metadata to see if leadership reassignment has finished or not.
+     */
+    private void waitForLeadershipTasksToFinish() {
+      List<ExecutionTask> finishedTasks = new ArrayList<>();
+      Set<Long> stoppedTaskIds = new HashSet<>();
+      Set<Long> deletedTaskIds = new HashSet<>();
+      Set<Long> deadTaskIds = new HashSet<>();
+      do {
+        // If there is no finished tasks, we need to check if anything is blocked.
+        maybeReexecuteLeadershipTasks();
+        Cluster cluster = getClusterForExecutionProgressCheck();
+
+        List<ExecutionTask> slowTasksToReport = new ArrayList<>();
+        boolean shouldReportSlowTasks = _time.milliseconds() - _lastSlowTaskReportingTimeMs > SLOW_TASK_ALERTING_BACKOFF_TIME_MS;
+        for (ExecutionTask task : inExecutionTasks()) {
+          TopicPartition tp = task.proposal().topicPartition();
+          if (_stopSignal.get() == FORCE_STOP_EXECUTION) {
+            // If the execution is force-stopped, the executor will mark all in progress tasks as dead, cleanup the
+            // relevant reassignment zNodes, and then bounce the controller (see ExecutionUtils#deleteZNodesToStopExecution).
```

Comment:
Is this comment correct? I thought CC no longer deletes the znode and moves controller in order to stop the execution.

Reply:
Yes, I believe it is correct.
Leadership reassignments in Kafka use `PreferredReplicaElection` znode; hence, potentially stuck leadership movements may require a force-stop. Though in practice, a stuck / time-consuming leadership movement is not common.

Note that `KIP-455` was about a new API for `inter-broker` replica reassignment tasks. Thus, its impact is limited to changes in starting, stopping, and listing them.
I cleaned up the `force-stop` logic to ensure that it only applies to leadership movements along with a deprecation message for replica reassignments  -- i.e. with `KIP-455` inter-broker replica reassignments no longer need force stop and intra-broker replica reassignments were already not using zNodes.

One side-effect of this change would be the loss of ability to bounce controller upon force-stopping inter/intra broker replica reassignments, which would have helped in case the controller itself is stuck for some reason. We can address that in a future PR by moving intentional controller bounce functionality to `admin` endpoint of cruise control.